File: 6장-선택/C-k번째 요소 찾기.py
'''
    KOREATECH 2021 Algorithm
    https://judge.koreatech.ac.kr/problem.php?id=1198
'''

import random
from sys import stdin

INF = -22000000000
testcase = int(stdin.readline().strip())


def RSelect(arr, left, right, target):
    arr = arr[left:right]
    if left == right:
        return arr[left]
    rand_idx = random.randint(0, len(arr) - 1)
    pivot = arr[rand_idx]
    t_left = [i for i in arr if i < pivot]
    t_mid = [i for i in arr if i == pivot]
    t_right = [i for i in arr if i > pivot]

    arr = t_left + t_mid + t_right

    mid_idx = len(t_left) + 1
    if mid_idx == target:
        return pivot
    elif mid_idx < target:
        return RSelect(arr, len(t_left) + len(t_mid), len(arr), target - mid_idx)  # 오른쪽
    elif mid_idx > target:
        return RSelect(arr, 0, mid_idx - 1, target)  # 왼쪽


while testcase:
    lenght, target = list(map(int, stdin.readline().strip().split()))
    arr = list(map(int, stdin.readline().strip().split()))

    print(RSelect(arr, 0, len(arr), target))
    testcase -= 1

File: 6장-선택/C-k번째 요소 찾기.java
/*
    KOREATECH 2021 Algorithm
    https://judge.koreatech.ac.kr/problem.php?id=1198
 */
import java.util.Scanner;
 
public class Main {
 
    public static int Partition(int[] arr, int n) {
        int p = arr[0];
        int i = 1;
        for (int j = 1; j < n; j++) {
            if (arr[j] < p) {
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                i++;
            }
        }
 
        int temp = arr[0];
        arr[0] = arr[i - 1];
        arr[i - 1] = temp;
 
        return i;
    }
 
    public static int RSelect(int[] arr, int n, int target) {
        int pivot_idx = (int) (Math.random() * n);
        int pivot = arr[pivot_idx];
 
        int temp = arr[0];
        arr[0] = arr[pivot_idx];
        arr[pivot_idx] = temp;
 
        int pos = Partition(arr, n);
 
        int[] new_arr = new int[0];
        int new_target = target;
        int new_n = 0;
 
        if (pos == target) {
            return pivot;
        } else if (pos < target) { // ¿À¸¥ÂÊ
            new_target -= pos;
            new_n = n - pos;
            new_arr = new int[new_n];
            for (int i = 0; i < new_n; i++) {
                new_arr[i] = arr[i + pos];
            }
        } else if (pos > target) { // ¿ÞÂÊ
            new_n = pos - 1;
            new_arr = new int[new_n];
            for (int i = 0; i < new_n; i++) {
                new_arr[i] = arr[i];
            }
        }
        return RSelect(new_arr, new_n, new_target);
 
    }
 
    public static void main(String[] args) {
        // TODO Auto-generated method stub
        Scanner sc = new Scanner(System.in);
        int testcase = sc.nextInt();
 
        while (testcase > 0) {
            int length = sc.nextInt();
            int target = sc.nextInt();
            int[] arr = new int[length];
 
            for (int i = 0; i < length; i++) {
                arr[i] = sc.nextInt();
            }
            System.out.println(RSelect(arr, length, target));
            testcase--;
        }
    }
}